### PR TITLE
[To rel/0.12] [IOTDB-1872 IOTDB-1902 IOTDB-1903] Fix data increases abnormally after IoTDB restarts

### PR DESCRIPTION
**Pull request by THUMarkLau**

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java [hunk 1]
```diff
@@ -695,6 +687,10 @@ private void recoverTsFiles(List<TsFileResource> tsFiles, boolean isSeq) throws
         logger.warn(
             "Skip TsFile: {} because of error in recover: ", tsFileResource.getTsFilePath(), e);
         continue;
+      } finally {
+        if (writer == null) {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java:691]:
```suggestion
        if (writer != null) {
```

Comment (THUMarkLau) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java]: resolve

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java [hunk 2]
```diff
@@ -1142,6 +1145,9 @@ public long selfCheck(
           this.position(),
           e.getMessage());
     }
+    if (loadLastChunkMetadata) {
+      chunkGroupMetadataList.add(new ChunkGroupMetadata(lastDeviceId, chunkMetadataList));
+    }
```
Comment (JackieTien97) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java:1150]: What if the last ChunkGroupMetadata has already been added before.

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java (hunk 2 above)
Comment (THUMarkLau) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java]: resolve

**Pull request merged**